File: examples/client/python/hami_client.py
#!/usr/bin/env python3
"""
HAMi Python Client

This is an example client library for interacting with HAMi (Heterogeneous AI Computing Virtualization Middleware).
HAMi integrates with Kubernetes to manage heterogeneous devices like GPUs, NPUs, etc.

The recommended way to use HAMi is through Kubernetes client libraries to manage pods and nodes.
This example shows how to:
1. Query HAMi device information from node annotations
2. Create pods with HAMi device requests
3. Monitor HAMi scheduling decisions

Requirements:
    pip install kubernetes

Usage:
    python hami_client.py
"""

from kubernetes import client, config
from typing import Dict, List, Optional
import json
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class HAMiClient:
    """Client for interacting with HAMi through Kubernetes API."""
    
    def __init__(self, kubeconfig: Optional[str] = None):
        """
        Initialize HAMi client.
        
        Args:
            kubeconfig: Path to kubeconfig file. If None, uses default config.
        """
        try:
            if kubeconfig:
                config.load_kube_config(config_file=kubeconfig)
            else:
                # Try in-cluster config first, then fall back to default kubeconfig
                try:
                    config.load_incluster_config()
                except config.ConfigException:
                    config.load_kube_config()
            
            self.core_v1 = client.CoreV1Api()
            logger.info("HAMi client initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize Kubernetes client: {e}")
            raise
    
    def list_gpu_nodes(self, label_selector: str = "gpu=on") -> List[Dict]:
        """
        List all nodes with HAMi GPU support.
        
        Args:
            label_selector: Label selector to filter nodes (default: "gpu=on")
            
        Returns:
            List of node information dictionaries
        """
        try:
            nodes = self.core_v1.list_node(label_selector=label_selector)
            result = []
            
            for node in nodes.items:
                node_info = {
                    "name": node.metadata.name,
                    "labels": node.metadata.labels,
                    "annotations": node.metadata.annotations or {},
                    "devices": self._parse_device_annotations(node.metadata.annotations or {})
                }
                result.append(node_info)
            
            logger.info(f"Found {len(result)} GPU nodes")
            return result
        except Exception as e:
            logger.error(f"Failed to list GPU nodes: {e}")
            raise
    
    def _parse_device_annotations(self, annotations: Dict[str, str]) -> Dict:
        """
        Parse HAMi device annotations from node.
        
        Args:
            annotations: Node annotations dictionary
            
        Returns:
            Dictionary of device information by device type
        """
        devices = {}
        
        for key, value in annotations.items():
            # Parse device registration annotations
            if key.startswith("hami.io/node-") and key.endswith("-register"):
                device_type = key.replace("hami.io/node-", "").replace("-register", "")
                devices[device_type] = self._parse_device_string(value)
            
            # Include handshake information
            if key.startswith("hami.io/node-handshake-"):
                device_type = key.replace("hami.io/node-handshake-", "")
                if device_type not in devices:
                    devices[device_type] = {}
                devices[device_type]["handshake"] = value
        
        return devices
    
    def _parse_device_string(self, device_str: str) -> List[Dict]:
        """
        Parse device registration string.
        
        Format: {Device UUID},{split count},{memory},{cores},{type},{numa},{healthy}:...
        
        Args:
            device_str: Device registration string
            
        Returns:
            List of device dictionaries
        """
        devices = []
        
        if not device_str or device_str.endswith(":"):
            device_str = device_str.rstrip(":")
        
        for device_info in device_str.split(":"):
            if not device_info:
                continue
            
            parts = device_info.split(",")
            if len(parts) >= 7:
                devices.append({
                    "uuid": parts[0],
                    "split_count": int(parts[1]),
                    "memory_mb": int(parts[2]),
                    "cores": int(parts[3]),
                    "type": parts[4],
                    "numa": int(parts[5]),
                    "healthy": parts[6].lower() == "true"
                })
        
        return devices
    
    def get_node_devices(self, node_name: str) -> Dict:
        """
        Get device information for a specific node.
        
        Args:
            node_name: Name of the node
            
        Returns:
            Dictionary of device information
        """
        try:
            node = self.core_v1.read_node(name=node_name)
            annotations = node.metadata.annotations or {}
            return self._parse_device_annotations(annotations)
        except Exception as e:
            logger.error(f"Failed to get devices for node {node_name}: {e}")
            raise
    
    def create_gpu_pod(self, 
                       pod_name: str,
                       namespace: str,
                       image: str,
                       gpu_count: int = 1,
                       gpu_memory_mb: Optional[int] = None,
                       gpu_cores: Optional[int] = None,
                       command: Optional[List[str]] = None) -> Dict:
        """
        Create a pod with GPU resources.
        
        Args:
            pod_name: Name of the pod
            namespace: Kubernetes namespace
            image: Container image
            gpu_count: Number of GPUs to request
            gpu_memory_mb: GPU memory in MB (optional)
            gpu_cores: GPU core percentage (optional)
            command: Container command (optional)
            
        Returns:
            Created pod object as dictionary
        """
        # Build resource limits
        limits = {
            "nvidia.com/gpu": str(gpu_count)
        }
        
        if gpu_memory_mb is not None:
            limits["nvidia.com/gpumem"] = str(gpu_memory_mb)
        
        if gpu_cores is not None:
            limits["nvidia.com/gpucores"] = str(gpu_cores)
        
        # Create pod specification
        container = client.V1Container(
            name=pod_name,
            image=image,
            command=command,
            resources=client.V1ResourceRequirements(
                limits=limits
            )
        )
        
        pod_spec = client.V1PodSpec(
            containers=[container],
            restart_policy="Never"
        )
        
        pod = client.V1Pod(
            api_version="v1",
            kind="Pod",
            metadata=client.V1ObjectMeta(name=pod_name),
            spec=pod_spec
        )
        
        try:
            created_pod = self.core_v1.create_namespaced_pod(
                namespace=namespace,
                body=pod
            )
            logger.info(f"Created pod {pod_name} in namespace {namespace}")
            return self._pod_to_dict(created_pod)
        except Exception as e:
            logger.error(f"Failed to create pod: {e}")
            raise
    
    def get_pod_scheduling_info(self, pod_name: str, namespace: str) -> Dict:
        """
        Get HAMi scheduling information for a pod.
        
        Args:
            pod_name: Name of the pod
            namespace: Kubernetes namespace
            
        Returns:
            Dictionary with scheduling information
        """
        try:
            pod = self.core_v1.read_namespaced_pod(name=pod_name, namespace=namespace)
            annotations = pod.metadata.annotations or {}
            
            scheduling_info = {
                "devices_to_allocate": annotations.get("hami.io/devices-to-allocate", ""),
                "node": annotations.get("hami.io/vgpu-node", ""),
                "schedule_time": annotations.get("hami.io/vgpu-time", ""),
                "pod_phase": pod.status.phase,
                "node_name": pod.spec.node_name
            }
            
            # Parse device allocation
            if scheduling_info["devices_to_allocate"]:
                scheduling_info["allocated_devices"] = self._parse_device_allocation(
                    scheduling_info["devices_to_allocate"]
                )
            
            return scheduling_info
        except Exception as e:
            logger.error(f"Failed to get pod scheduling info: {e}")
            raise
    
    def _parse_device_allocation(self, allocation_str: str) -> List[Dict]:
        """
        Parse device allocation string.
        
        Format: {device UUID},{type},{memory},{cores}:...
        
        Args:
            allocation_str: Device allocation string
            
        Returns:
            List of allocated device dictionaries
        """
        devices = []
        
        if not allocation_str or allocation_str.endswith(":"):
            allocation_str = allocation_str.rstrip(":")
        
        for device_info in allocation_str.split(":"):
            if not device_info:
                continue
            
            parts = device_info.split(",")
            if len(parts) >= 4:
                devices.append({
                    "uuid": parts[0],
                    "type": parts[1],
                    "memory_mb": int(parts[2]),
                    "cores": int(parts[3])
                })
        
        return devices
    
    def _pod_to_dict(self, pod) -> Dict:
        """Convert pod object to dictionary."""
        return {
            "name": pod.metadata.name,
            "namespace": pod.metadata.namespace,
            "uid": pod.metadata.uid,
            "phase": pod.status.phase,
            "node_name": pod.spec.node_name
        }
    
    def delete_pod(self, pod_name: str, namespace: str) -> None:
        """
        Delete a pod.
        
        Args:
            pod_name: Name of the pod
            namespace: Kubernetes namespace
        """
        try:
            self.core_v1.delete_namespaced_pod(
                name=pod_name,
                namespace=namespace
            )
            logger.info(f"Deleted pod {pod_name} from namespace {namespace}")
        except Exception as e:
            logger.error(f"Failed to delete pod: {e}")
            raise


def main():
    """Example usage of HAMi client."""
    # Initialize client
    hami = HAMiClient()
    
    # List GPU nodes
    print("\n=== GPU Nodes ===")
    nodes = hami.list_gpu_nodes()
    for node in nodes:
        print(f"\nNode: {node['name']}")
        print(f"Devices: {json.dumps(node['devices'], indent=2)}")
    
    # Example: Create a GPU pod
    print("\n=== Creating GPU Pod ===")
    try:
        pod_info = hami.create_gpu_pod(
            pod_name="hami-example-pod",
            namespace="default",
            image="nvidia/cuda:11.0-base",
            gpu_count=1,
            gpu_memory_mb=3000,
            gpu_cores=50,
            command=["nvidia-smi"]
        )
        print(f"Created pod: {json.dumps(pod_info, indent=2)}")
        
        # Get scheduling info
        import time
        time.sleep(2)  # Wait for scheduling
        
        print("\n=== Pod Scheduling Info ===")
        scheduling_info = hami.get_pod_scheduling_info("hami-example-pod", "default")
        print(json.dumps(scheduling_info, indent=2))
        
        # Clean up
        print("\n=== Cleaning Up ===")
        hami.delete_pod("hami-example-pod", "default")
        
    except Exception as e:
        logger.error(f"Example failed: {e}")


if __name__ == "__main__":
    main()

File: examples/client/java/HAMiClient.java
/**
 * HAMi Java Client Example
 * 
 * This is an example client library for interacting with HAMi 
 * (Heterogeneous AI Computing Virtualization Middleware).
 * 
 * HAMi integrates with Kubernetes to manage heterogeneous devices like GPUs, NPUs, etc.
 * 
 * Requirements:
 * - Kubernetes Java Client: io.kubernetes:client-java
 * - Maven or Gradle for dependency management
 * 
 * Usage:
 *     javac -cp kubernetes-client.jar HAMiClient.java
 *     java -cp .:kubernetes-client.jar HAMiClient
 */

package io.hami.client;

import io.kubernetes.client.openapi.ApiClient;
import io.kubernetes.client.openapi.ApiException;
import io.kubernetes.client.openapi.Configuration;
import io.kubernetes.client.openapi.apis.CoreV1Api;
import io.kubernetes.client.openapi.models.*;
import io.kubernetes.client.util.Config;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Client for interacting with HAMi through Kubernetes API.
 */
public class HAMiClient {
    
    private final CoreV1Api coreV1Api;
    
    /**
     * Initialize HAMi client with default configuration.
     * 
     * @throws IOException if kubeconfig cannot be loaded
     */
    public HAMiClient() throws IOException {
        ApiClient client = Config.defaultClient();
        Configuration.setDefaultApiClient(client);
        this.coreV1Api = new CoreV1Api();
    }
    
    /**
     * Initialize HAMi client with custom API client.
     * 
     * @param apiClient Custom Kubernetes API client
     */
    public HAMiClient(ApiClient apiClient) {
        Configuration.setDefaultApiClient(apiClient);
        this.coreV1Api = new CoreV1Api();
    }
    
    /**
     * List all nodes with HAMi GPU support.
     * 
     * @param labelSelector Label selector to filter nodes (default: "gpu=on")
     * @return List of node information
     * @throws ApiException if API call fails
     */
    public List<NodeInfo> listGPUNodes(String labelSelector) throws ApiException {
        if (labelSelector == null) {
            labelSelector = "gpu=on";
        }
        
        V1NodeList nodeList = coreV1Api.listNode(
            null, null, null, null, labelSelector, 
            null, null, null, null, null, null
        );
        
        List<NodeInfo> result = new ArrayList<>();
        for (V1Node node : nodeList.getItems()) {
            NodeInfo nodeInfo = new NodeInfo();
            nodeInfo.name = node.getMetadata().getName();
            nodeInfo.labels = node.getMetadata().getLabels();
            nodeInfo.annotations = node.getMetadata().getAnnotations();
            nodeInfo.devices = parseDeviceAnnotations(nodeInfo.annotations);
            result.add(nodeInfo);
        }
        
        System.out.println("Found " + result.size() + " GPU nodes");
        return result;
    }
    
    /**
     * Parse HAMi device annotations from node.
     * 
     * @param annotations Node annotations
     * @return Map of device information by device type
     */
    private Map<String, DeviceInfo> parseDeviceAnnotations(Map<String, String> annotations) {
        Map<String, DeviceInfo> devices = new HashMap<>();
        
        if (annotations == null) {
            return devices;
        }
        
        for (Map.Entry<String, String> entry : annotations.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            
            // Parse device registration annotations
            if (key.startsWith("hami.io/node-") && key.endsWith("-register")) {
                String deviceType = key.replace("hami.io/node-", "")
                                      .replace("-register", "");
                DeviceInfo deviceInfo = new DeviceInfo();
                deviceInfo.type = deviceType;
                deviceInfo.devices = parseDeviceString(value);
                devices.put(deviceType, deviceInfo);
            }
            
            // Include handshake information
            if (key.startsWith("hami.io/node-handshake-")) {
                String deviceType = key.replace("hami.io/node-handshake-", "");
                DeviceInfo deviceInfo = devices.getOrDefault(deviceType, new DeviceInfo());
                deviceInfo.handshake = value;
                devices.put(deviceType, deviceInfo);
            }
        }
        
        return devices;
    }
    
    /**
     * Parse device registration string.
     * Format: {Device UUID},{split count},{memory},{cores},{type},{numa},{healthy}:...
     * 
     * @param deviceStr Device registration string
     * @return List of device details
     */
    private List<Device> parseDeviceString(String deviceStr) {
        List<Device> devices = new ArrayList<>();
        
        if (deviceStr == null || deviceStr.isEmpty()) {
            return devices;
        }
        
        deviceStr = deviceStr.replaceAll(":+$", "");
        String[] deviceInfos = deviceStr.split(":");
        
        for (String deviceInfo : deviceInfos) {
            if (deviceInfo.isEmpty()) {
                continue;
            }
            
            String[] parts = deviceInfo.split(",");
            if (parts.length >= 7) {
                Device device = new Device();
                device.uuid = parts[0];
                device.splitCount = Integer.parseInt(parts[1]);
                device.memoryMB = Integer.parseInt(parts[2]);
                device.cores = Integer.parseInt(parts[3]);
                device.type = parts[4];
                device.numa = Integer.parseInt(parts[5]);
                device.healthy = Boolean.parseBoolean(parts[6]);
                devices.add(device);
            }
        }
        
        return devices;
    }
    
    /**
     * Get device information for a specific node.
     * 
     * @param nodeName Name of the node
     * @return Map of device information
     * @throws ApiException if API call fails
     */
    public Map<String, DeviceInfo> getNodeDevices(String nodeName) throws ApiException {
        V1Node node = coreV1Api.readNode(nodeName, null);
        Map<String, String> annotations = node.getMetadata().getAnnotations();
        return parseDeviceAnnotations(annotations);
    }
    
    /**
     * Create a pod with GPU resources.
     * 
     * @param podName Name of the pod
     * @param namespace Kubernetes namespace
     * @param image Container image
     * @param gpuCount Number of GPUs to request
     * @param gpuMemoryMB GPU memory in MB (optional, use null to skip)
     * @param gpuCores GPU core percentage (optional, use null to skip)
     * @param command Container command (optional)
     * @return Created pod information
     * @throws ApiException if API call fails
     */
    public PodInfo createGPUPod(String podName, String namespace, String image,
                                int gpuCount, Integer gpuMemoryMB, 
                                Integer gpuCores, List<String> command) throws ApiException {
        // Build resource limits
        Map<String, io.kubernetes.client.custom.Quantity> limits = new HashMap<>();
        limits.put("nvidia.com/gpu", new io.kubernetes.client.custom.Quantity(String.valueOf(gpuCount)));
        
        if (gpuMemoryMB != null) {
            limits.put("nvidia.com/gpumem", new io.kubernetes.client.custom.Quantity(String.valueOf(gpuMemoryMB)));
        }
        
        if (gpuCores != null) {
            limits.put("nvidia.com/gpucores", new io.kubernetes.client.custom.Quantity(String.valueOf(gpuCores)));
        }
        
        // Create container
        V1Container container = new V1Container()
            .name(podName)
            .image(image)
            .command(command)
            .resources(new V1ResourceRequirements().limits(limits));
        
        // Create pod spec
        V1PodSpec podSpec = new V1PodSpec()
            .containers(Collections.singletonList(container))
            .restartPolicy("Never");
        
        // Create pod
        V1Pod pod = new V1Pod()
            .apiVersion("v1")
            .kind("Pod")
            .metadata(new V1ObjectMeta().name(podName))
            .spec(podSpec);
        
        V1Pod createdPod = coreV1Api.createNamespacedPod(namespace, pod, null, null, null, null);
        System.out.println("Created pod " + podName + " in namespace " + namespace);
        
        return podToInfo(createdPod);
    }
    
    /**
     * Get HAMi scheduling information for a pod.
     * 
     * @param podName Name of the pod
     * @param namespace Kubernetes namespace
     * @return Scheduling information
     * @throws ApiException if API call fails
     */
    public SchedulingInfo getPodSchedulingInfo(String podName, String namespace) throws ApiException {
        V1Pod pod = coreV1Api.readNamespacedPod(podName, namespace, null);
        Map<String, String> annotations = pod.getMetadata().getAnnotations();
        
        if (annotations == null) {
            annotations = new HashMap<>();
        }
        
        SchedulingInfo info = new SchedulingInfo();
        info.devicesToAllocate = annotations.getOrDefault("hami.io/devices-to-allocate", "");
        info.node = annotations.getOrDefault("hami.io/vgpu-node", "");
        info.scheduleTime = annotations.getOrDefault("hami.io/vgpu-time", "");
        info.podPhase = pod.getStatus().getPhase();
        info.nodeName = pod.getSpec().getNodeName();
        
        // Parse device allocation
        if (!info.devicesToAllocate.isEmpty()) {
            info.allocatedDevices = parseDeviceAllocation(info.devicesToAllocate);
        }
        
        return info;
    }
    
    /**
     * Parse device allocation string.
     * Format: {device UUID},{type},{memory},{cores}:...
     * 
     * @param allocationStr Device allocation string
     * @return List of allocated devices
     */
    private List<AllocatedDevice> parseDeviceAllocation(String allocationStr) {
        List<AllocatedDevice> devices = new ArrayList<>();
        
        if (allocationStr == null || allocationStr.isEmpty()) {
            return devices;
        }
        
        allocationStr = allocationStr.replaceAll(":+$", "");
        String[] deviceInfos = allocationStr.split(":");
        
        for (String deviceInfo : deviceInfos) {
            if (deviceInfo.isEmpty()) {
                continue;
            }
            
            String[] parts = deviceInfo.split(",");
            if (parts.length >= 4) {
                AllocatedDevice device = new AllocatedDevice();
                device.uuid = parts[0];
                device.type = parts[1];
                device.memoryMB = Integer.parseInt(parts[2]);
                device.cores = Integer.parseInt(parts[3]);
                devices.add(device);
            }
        }
        
        return devices;
    }
    
    /**
     * Delete a pod.
     * 
     * @param podName Name of the pod
     * @param namespace Kubernetes namespace
     * @throws ApiException if API call fails
     */
    public void deletePod(String podName, String namespace) throws ApiException {
        coreV1Api.deleteNamespacedPod(
            podName, namespace, null, null, 
            null, null, null, null
        );
        System.out.println("Deleted pod " + podName + " from namespace " + namespace);
    }
    
    /**
     * Convert pod object to info.
     */
    private PodInfo podToInfo(V1Pod pod) {
        PodInfo info = new PodInfo();
        info.name = pod.getMetadata().getName();
        info.namespace = pod.getMetadata().getNamespace();
        info.uid = pod.getMetadata().getUid();
        info.phase = pod.getStatus().getPhase();
        info.nodeName = pod.getSpec().getNodeName();
        return info;
    }
    
    // Data classes
    
    public static class NodeInfo {
        public String name;
        public Map<String, String> labels;
        public Map<String, String> annotations;
        public Map<String, DeviceInfo> devices;
        
        @Override
        public String toString() {
            return "NodeInfo{name='" + name + "', devices=" + devices + "}";
        }
    }
    
    public static class DeviceInfo {
        public String type;
        public String handshake;
        public List<Device> devices;
        
        @Override
        public String toString() {
            return "DeviceInfo{type='" + type + "', devices=" + devices + "}";
        }
    }
    
    public static class Device {
        public String uuid;
        public int splitCount;
        public int memoryMB;
        public int cores;
        public String type;
        public int numa;
        public boolean healthy;
        
        @Override
        public String toString() {
            return "Device{uuid='" + uuid + "', memory=" + memoryMB + "MB, cores=" + cores + "%, healthy=" + healthy + "}";
        }
    }
    
    public static class PodInfo {
        public String name;
        public String namespace;
        public String uid;
        public String phase;
        public String nodeName;
        
        @Override
        public String toString() {
            return "PodInfo{name='" + name + "', namespace='" + namespace + "', phase='" + phase + "'}";
        }
    }
    
    public static class SchedulingInfo {
        public String devicesToAllocate;
        public String node;
        public String scheduleTime;
        public String podPhase;
        public String nodeName;
        public List<AllocatedDevice> allocatedDevices;
        
        @Override
        public String toString() {
            return "SchedulingInfo{node='" + node + "', allocatedDevices=" + allocatedDevices + "}";
        }
    }
    
    public static class AllocatedDevice {
        public String uuid;
        public String type;
        public int memoryMB;
        public int cores;
        
        @Override
        public String toString() {
            return "AllocatedDevice{uuid='" + uuid + "', type='" + type + "', memory=" + memoryMB + "MB, cores=" + cores + "%}";
        }
    }
    
    /**
     * Example usage of HAMi client.
     */
    public static void main(String[] args) {
        try {
            // Initialize client
            HAMiClient hami = new HAMiClient();
            
            // List GPU nodes
            System.out.println("\n=== GPU Nodes ===");
            List<NodeInfo> nodes = hami.listGPUNodes("gpu=on");
            for (NodeInfo node : nodes) {
                System.out.println(node);
            }
            
            // Example: Create a GPU pod
            System.out.println("\n=== Creating GPU Pod ===");
            PodInfo podInfo = hami.createGPUPod(
                "hami-example-pod",
                "default",
                "nvidia/cuda:11.0-base",
                1,           // GPU count
                3000,        // GPU memory MB
                50,          // GPU cores %
                Arrays.asList("nvidia-smi")
            );
            System.out.println(podInfo);
            
            // Wait for scheduling
            Thread.sleep(2000);
            
            // Get scheduling info
            System.out.println("\n=== Pod Scheduling Info ===");
            SchedulingInfo schedulingInfo = hami.getPodSchedulingInfo("hami-example-pod", "default");
            System.out.println(schedulingInfo);
            
            // Clean up
            System.out.println("\n=== Cleaning Up ===");
            hami.deletePod("hami-example-pod", "default");
            
        } catch (Exception e) {
            System.err.println("Example failed: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
